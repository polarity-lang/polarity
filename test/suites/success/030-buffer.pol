data Nat { Z, S(n: Nat) }

data Bool { T, F }

codata Pair(a b: Type) {
    Pair(a, b).proj1(a b: Type): a,
    Pair(a, b).proj2(a b: Type): b
}

data Vec(n: Nat) {
    Nil: Vec(Z),
    Cons(n: Nat, x: Bool, xs: Vec(n)): Vec(S(n))
}

def Vec(S(n)).head(n: Nat): Bool {
    Cons(n', x, xs) => x,
    Nil absurd
}

def Vec(S(n)).tail(n: Nat): Vec(n) {
    Cons(n', x, xs) => xs,
    Nil absurd
}

codata Buffer(n: Nat) {
    Buffer(S(n)).read(n: Nat): Pair(Bool, Buffer(n))
}

codef Empty: Buffer(Z) { .read(n) absurd }

codef FromVec(n: Nat, xs: Vec(n)): Buffer(n) {
    .read(n') =>
        comatch {
            .proj1(_, _) => xs.head(n'),
            .proj2(_, _) => FromVec(n', xs.tail(n'))
        }
}
