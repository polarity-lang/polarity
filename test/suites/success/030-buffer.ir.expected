def .head(n) {
    Cons(n', x, xs) => x,
    Nil absurd
}

def .tail(n) {
    Cons(n', x, xs) => xs,
    Nil absurd
}

codef Empty { .read(n) absurd }

codef FromVec(n, xs) {
    .read(n') =>
        comatch {
            .proj1(x, x0) => xs.head(n'),
            .proj2(x, x0) => FromVec(n', xs.tail(n'))
        }
}
